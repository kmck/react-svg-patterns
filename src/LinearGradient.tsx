import * as React from 'react';

import {
  GradientColor,
  SpreadMethod,
  UnprocessedGradientStops,
} from './types';

import {
  DEGREES_TO_RADIANS,
  SPREAD_METHOD_PAD,
} from './constants'
import { processStops } from './utils';

import GradientStop from './GradientStop';

interface Props {
  angle?: number,
  from?: GradientColor,
  id: string,
  spreadMethod?: SpreadMethod,
  scale?: number,
  stops?: UnprocessedGradientStops,
  to?: GradientColor
};

const LinearGradient = ({
  angle = 0,
  from = 'inherit',
  id,
  scale = 1,
  spreadMethod = SPREAD_METHOD_PAD,
  stops,
  to = 'inherit'
}: Props) => {
  const gradientStops = React.useMemo(
    () => processStops(stops || ([from, to])),
    [from, to, stops],
  );

  // 0 degrees goes bottom-to-top, 90 degrees goes left-to-right, etc.
  const { x1, x2, y1, y2 } = React.useMemo(() => {
    const radians = angle * DEGREES_TO_RADIANS;
    const x = scale * Math.sin(radians);
    const y = scale * Math.cos(radians);
    const xl = Math.max(-x, 0);
    const xr = Math.max(0, x);
    const yl = Math.max(y, 0);
    const yr = Math.max(0, -y);

    // Offset to keep the gradient centered
    const xRange = Math.abs(xr - xl);
    const yRange = Math.abs(yr - yl);
    const xo = 0.5 * (1 - xRange);
    const yo = 0.5 * (1 - yRange);

    return {
      x1: xl + xo,
      x2: xr + xo,
      y1: yl + yo,
      y2: yr + yo,
    };
  }, [angle]);

  return (
    <linearGradient
      id={id}
      x1={x1}
      x2={x2}
      y1={y1}
      y2={y2}
      spreadMethod={spreadMethod}
    >
      {gradientStops.map((stopProps, key) => (
        <GradientStop
          key={stopProps.offset != null ? stopProps.offset : key}
          {...stopProps}
        />
      ))}
    </linearGradient>
  );
};

export default React.memo(LinearGradient);
