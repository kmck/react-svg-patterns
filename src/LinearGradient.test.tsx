import * as React from 'react';
import renderer from 'react-test-renderer';

import LinearGradient from './LinearGradient';

describe('LineraGradient', () => {
  it('renders a linear gradient', () => {
    const tree = renderer.create(
      <LinearGradient id="test" />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a linear gradient using from/to', () => {
    const tree = renderer.create(
      <LinearGradient id="test" from="#ff0000" to="#0000ff" />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a linear gradient using unprocessed stops', () => {
    const tree = renderer.create(
      <LinearGradient id="test" stops={['#ff0000', '#00ff00', '#0000ff']} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a linear gradient using preprocessed stops', () => {
    const tree = renderer.create(
      <LinearGradient id="test" stops={[{ offset: 0.25, color: '#000' }, { offset: 0.75, color: '#fff' }]} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a linear gradient at an angle', () => {
    const tree = renderer.create(
      <LinearGradient id="test" from="#fff" to="#000" angle={45} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
    expect(tree).not.toEqual(renderer.create(
      <LinearGradient id="test" from="#fff" to="#000" />
    ).toJSON());
  });

  it('renders a linear gradient with spreadMethod', () => {
    const tree = renderer.create(
      <>
        <LinearGradient id="test1" from="#fff" to="#000" spreadMethod="pad" />
        <LinearGradient id="test2" from="#fff" to="#000" spreadMethod="reflect" />
        <LinearGradient id="test3" from="#fff" to="#000" spreadMethod="repeat" />
      </>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a linear gradient with scale', () => {
    const tree = renderer.create(
      <LinearGradient id="test" from="#fff" to="#000" scale={2} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
    expect(tree).not.toEqual(renderer.create(
      <LinearGradient id="test" from="#fff" to="#000" />
    ).toJSON());
  });
});
