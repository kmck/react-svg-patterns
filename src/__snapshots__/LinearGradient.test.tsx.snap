// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineraGradient renders a linear gradient 1`] = `
<linearGradient
  id="test"
  spreadMethod="pad"
  x1={0.5}
  x2={0.5}
  y1={1}
  y2={0}
>
  <stop
    offset="0%"
    stopColor="inherit"
  />
  <stop
    offset="100%"
    stopColor="inherit"
  />
</linearGradient>
`;

exports[`LineraGradient renders a linear gradient at an angle 1`] = `
<linearGradient
  id="test"
  spreadMethod="pad"
  x1={0.14644660940672627}
  x2={0.8535533905932737}
  y1={0.8535533905932737}
  y2={0.1464466094067262}
>
  <stop
    offset="0%"
    stopColor="#fff"
  />
  <stop
    offset="100%"
    stopColor="#000"
  />
</linearGradient>
`;

exports[`LineraGradient renders a linear gradient using from/to 1`] = `
<linearGradient
  id="test"
  spreadMethod="pad"
  x1={0.5}
  x2={0.5}
  y1={1}
  y2={0}
>
  <stop
    offset="0%"
    stopColor="#ff0000"
  />
  <stop
    offset="100%"
    stopColor="#0000ff"
  />
</linearGradient>
`;

exports[`LineraGradient renders a linear gradient using preprocessed stops 1`] = `
<linearGradient
  id="test"
  spreadMethod="pad"
  x1={0.5}
  x2={0.5}
  y1={1}
  y2={0}
>
  <stop
    offset="25%"
    stopColor="#000"
  />
  <stop
    offset="75%"
    stopColor="#fff"
  />
</linearGradient>
`;

exports[`LineraGradient renders a linear gradient using unprocessed stops 1`] = `
<linearGradient
  id="test"
  spreadMethod="pad"
  x1={0.5}
  x2={0.5}
  y1={1}
  y2={0}
>
  <stop
    offset="0%"
    stopColor="#ff0000"
  />
  <stop
    offset="50%"
    stopColor="#00ff00"
  />
  <stop
    offset="100%"
    stopColor="#0000ff"
  />
</linearGradient>
`;

exports[`LineraGradient renders a linear gradient with scale 1`] = `
<linearGradient
  id="test"
  spreadMethod="pad"
  x1={0.5}
  x2={0.5}
  y1={1.5}
  y2={-0.5}
>
  <stop
    offset="0%"
    stopColor="#fff"
  />
  <stop
    offset="100%"
    stopColor="#000"
  />
</linearGradient>
`;

exports[`LineraGradient renders a linear gradient with spreadMethod 1`] = `
Array [
  <linearGradient
    id="test1"
    spreadMethod="pad"
    x1={0.5}
    x2={0.5}
    y1={1}
    y2={0}
  >
    <stop
      offset="0%"
      stopColor="#fff"
    />
    <stop
      offset="100%"
      stopColor="#000"
    />
  </linearGradient>,
  <linearGradient
    id="test2"
    spreadMethod="reflect"
    x1={0.5}
    x2={0.5}
    y1={1}
    y2={0}
  >
    <stop
      offset="0%"
      stopColor="#fff"
    />
    <stop
      offset="100%"
      stopColor="#000"
    />
  </linearGradient>,
  <linearGradient
    id="test3"
    spreadMethod="repeat"
    x1={0.5}
    x2={0.5}
    y1={1}
    y2={0}
  >
    <stop
      offset="0%"
      stopColor="#fff"
    />
    <stop
      offset="100%"
      stopColor="#000"
    />
  </linearGradient>,
]
`;
